pkg/cmd/server/origin_master.go

```diff
@@ -210,21 +217,84 @@ func (cfg Config) newCA() (*crypto.CA, error) {
 	return ca, nil
 }
 
-func (cfg Config) GetClientCAs() (*x509.CertPool, error) {
-	ca, err := cfg.newCA()
+// GetAPIClientCertPool returns the cert pool used to validate client certificates to the API server
+func (cfg Config) GetAPIClientCertCAPool() (*x509.CertPool, error) {
+	certs, err := cfg.getAPIClientCertCAs()
 	if err != nil {
 		return nil, err
 	}
+	roots := x509.NewCertPool()
+	for _, root := range certs {
+		roots.AddCert(root)
+	}
+	return roots, nil
+}
 
-	// Save cert roots
+// GetClientCertCAPool returns a cert pool containing all client CAs (API and OAuth)
+func (cfg Config) GetClientCertCAPool() (*x509.CertPool, error) {
 	roots := x509.NewCertPool()
-	for _, root := range ca.Config.Roots {
+
+	// Add CAs for OAuth
+	certs, err := cfg.getOAuthClientCertCAs()
+	if err != nil {
+		return nil, err
+	}
+	for _, root := range certs {
+		roots.AddCert(root)
+	}
+
+	// Add CAs for API
+	certs, err = cfg.getAPIClientCertCAs()
+	if err != nil {
+		return nil, err
+	}
+	for _, root := range certs {
 		roots.AddCert(root)
 	}
 
 	return roots, nil
 }
 
+// GetAPIServerCertPool returns the cert pool containing the roots for the API server cert
+func (cfg Config) GetAPIServerCertCAPool() (*x509.CertPool, error) {
+	certs, err := cfg.getAPIClientCertCAs()
+	if err != nil {
+		return nil, err
+	}
+	roots := x509.NewCertPool()
+	for _, root := range certs {
+		roots.AddCert(root)
+	}
+	return roots, nil
+}
+
+func (cfg Config) getOAuthClientCertCAs() ([]*x509.Certificate, error) {
+	auth, err := cfg.BuildAuthConfig()
```

Comment:
Pulled out the `OPENSHIFT_OAUTH_REQUEST_HEADER_CA_FILE` lookup into `GetOAuthRequestHeaderCAFile`